File: ios/StudipadawanWidgets/Models/ScheduleResponse.swift
//
//  ScheduleResponse.swift
//  StudipadawanWidgetsExtension
//
//  Created by Finn Ebeling on 02.06.23.
//

import Foundation

struct ScheduleResponse: Codable {
    let data: [Data]
    
    struct Data: Codable {
        let type: String
        let id: String
        let attributes: Attribute
        
        struct Attribute: Codable {
            let title: String
            let description: String?
            let start: String
            let end: String
            let weekday: Weekday
            let locations: [String]?
            let recurrence: Recurrence?
            
            enum Weekday: Int, Codable {
                case monday = 1
                case tuesday, wednesday, thursday, friday, saturday, sunday
            }
            
            struct Recurrence: Codable {
                let freq: String
                let interval: Int
                let firstOccurence: Date
                let lastOccurence: Date
                let excludedDates: [Date]?
                
                private enum CodingKeys: String, CodingKey {
                    case freq = "FREQ"
                    case interval = "INTERVAL"
                    case firstOccurence = "DTSTART"
                    case lastOccurence = "UNTIL"
                    case excludedDates = "EXDATES"
                }
            }
        }
    }
}

File: packages/studip_api_client/lib/src/models/schedule/schedule_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'schedule_response.g.dart';

@JsonSerializable()
class ScheduleListResponse {
  ScheduleListResponse({required this.scheduleEntries});

  factory ScheduleListResponse.fromJson(Map<String, dynamic> json) =>
      _$ScheduleListResponseFromJson(json);

  @JsonKey(name: 'data')
  final List<ScheduleResponseItem> scheduleEntries;
}

@JsonSerializable()
class ScheduleResponseItem {
  ScheduleResponseItem({
    required this.type,
    required this.id,
    required this.attributes,
    required this.relationships,
  });

  factory ScheduleResponseItem.fromJson(Map<String, dynamic> json) =>
      _$ScheduleResponseItemFromJson(json);
  final String type;
  final String id;
  final ScheduleResponseItemAttributes attributes;
  final ScheduleResponseItemRelationships relationships;
}

// --- Attributes ---

@JsonSerializable()
class ScheduleResponseItemAttributes {
  ScheduleResponseItemAttributes({
    required this.title,
    required this.description,
    required this.start,
    required this.end,
    required this.weekday,
    required this.locations,
    required this.recurrence,
  });

  factory ScheduleResponseItemAttributes.fromJson(Map<String, dynamic> json) =>
      _$ScheduleResponseItemAttributesFromJson(json);

  final String title;
  final String? description;
  final String start;
  final String end;
  final int weekday;
  final List<String>? locations;
  final ScheduleResponseItemAttributesRecurrence? recurrence;
}

@JsonSerializable()
class ScheduleResponseItemAttributesRecurrence {
  ScheduleResponseItemAttributesRecurrence({
    required this.freq,
    required this.interval,
    required this.firstOccurrenceDateString,
    required this.lastOccurrenceDateString,
    required this.excludedDates,
  });

  factory ScheduleResponseItemAttributesRecurrence.fromJson(
    Map<String, dynamic> json,
  ) =>
      _$ScheduleResponseItemAttributesRecurrenceFromJson(json);
  @JsonKey(name: 'FREQ')
  final String freq;

  @JsonKey(name: 'INTERVAL')
  final int interval;

  @JsonKey(name: 'DTSTART')
  final String firstOccurrenceDateString;

  @JsonKey(name: 'UNTIL')
  final String lastOccurrenceDateString;

  @JsonKey(name: 'EXDATES')
  final List<String>? excludedDates;
}

// --- Relationshipss ---

@JsonSerializable()
class ScheduleResponseItemRelationships {
  ScheduleResponseItemRelationships({required this.owner});

  factory ScheduleResponseItemRelationships.fromJson(
    Map<String, dynamic> json,
  ) =>
      _$ScheduleResponseItemRelationshipsFromJson(json);

  final ScheduleResponseItemRelationshipsOwner owner;
}

@JsonSerializable()
class ScheduleResponseItemRelationshipsOwner {
  ScheduleResponseItemRelationshipsOwner({required this.data});

  factory ScheduleResponseItemRelationshipsOwner.fromJson(
    Map<String, dynamic> json,
  ) =>
      _$ScheduleResponseItemRelationshipsOwnerFromJson(json);

  final ScheduleResponseItemRelationshipsOwnerData data;
}

@JsonSerializable()
class ScheduleResponseItemRelationshipsOwnerData {
  ScheduleResponseItemRelationshipsOwnerData({
    required this.type,
    required this.id,
  });

  factory ScheduleResponseItemRelationshipsOwnerData.fromJson(
    Map<String, dynamic> json,
  ) =>
      _$ScheduleResponseItemRelationshipsOwnerDataFromJson(json);

  final String type;
  final String id;
}
